src/app/frontend/common/auth/service.js

```diff
@@ -135,21 +137,28 @@ export class AuthService {
    */
   isLoggedIn(transition) {
     let deferred = this.q_.defer();
-
-    // Skip log in check if user is going to login page already or has chosen to skip it.
-    if (!this.isLoginPageEnabled() || transition.to().name === loginState ||
-        transition.to().name === 'internalerror') {
-      deferred.resolve(true);
-      return deferred.promise;
-    }
-
     this.getLoginStatus().then(
         (/** @type {!backendApi.LoginStatus} */ loginStatus) => {
-          if (loginStatus.headerPresent || loginStatus.tokenPresent) {
+          // Disable entering login page from HTTP.
+          if (transition.to().name === loginState && !loginStatus.httpsMode) {
+            deferred.resolve(this.state_.target(overviewState));
+            return;
+          }
+
+          // Do not redirect if user is going to login page already or has chosen to skip it.
+          if (!this.isLoginPageEnabled() || transition.to().name === loginState ||
+              transition.to().name === 'internalerror') {
+            deferred.resolve(true);
+            return deferred.promise;
+          }
+
+          // Do not redirect to login if already authenticated or using HTTP.
+          if (loginStatus.headerPresent || loginStatus.tokenPresent || !loginStatus.httpsMode) {
```

Comment:
This won't actually prevent user from entering login page. He can still go to login page by changing url. I'd merge this check with the first one.